barretenberg/cpp/src/barretenberg/vm2/common/uint1.hpp

```diff
@@ -0,0 +1,75 @@
+#pragma once
+
+#include <cstdint>
+#include <type_traits>
+
+namespace bb::avm2 {
+
+/**
+ * @brief A 1-bit unsigned integer type.
+ *
+ * This type is used to represent boolean values in a way that's compatible
+ * with the TaggedValue system. It behaves like a regular integer type but
+ * only allows values 0 and 1.
+ */
+class uint1_t {
+  public:
+    uint1_t(const uint1_t& other) = default;
+    uint1_t& operator=(const uint1_t& other) = default;
+
+    // Default constructor initializes to 0
+    constexpr uint1_t() noexcept
+        : value_(false)
+    {}
+
+    // Constructor from bool
+    constexpr explicit uint1_t(bool b) noexcept
+        : value_(b)
+    {}
+
+    // Constructor from integral types, enforcing 0 or 1 values
+    template <typename T, typename = std::enable_if_t<std::is_integral_v<T>>>
+    constexpr explicit uint1_t(T v) noexcept
+        : value_(v != 0)
+    {}
+
+    // Arithmetic operators
+    constexpr uint1_t operator+(const uint1_t& other) const noexcept { return *this ^ other; }
+    constexpr uint1_t operator-(const uint1_t& other) const noexcept { return *this + other; }
+    constexpr uint1_t operator*(const uint1_t& other) const noexcept { return *this & other; }
+    constexpr uint1_t operator/(const uint1_t& other) const noexcept { return uint1_t(value() / other.value()); }
+    constexpr uint1_t operator-() const noexcept { return uint1_t(!value_); }
+
+    // Bitwise operators
+    constexpr uint1_t operator&(const uint1_t& other) const noexcept { return uint1_t(value_ && other.value_); }
+    constexpr uint1_t operator|(const uint1_t& other) const noexcept { return uint1_t(value_ || other.value_); }
+    constexpr uint1_t operator^(const uint1_t& other) const noexcept { return uint1_t(value_ != other.value_); }
+    constexpr uint1_t operator~() const noexcept { return uint1_t(!value_); }
+
+    // Shifts are a bit special.
```

Comment:
i think we can do both shifts with this : `return (*this - (*this * other));`

====
Given `c = a >> b` if `b == 1`(or true), then `c == 0`, otherwise `c == a`

So given `a >> b` we should be able to encode `b ? 0 : a` as ==> `c = a - ab`

Reply:
Done thanks! Tests pass.